eventsourced/src/lib.rs

```diff
@@ -66,204 +55,111 @@ use tokio::{
-/// which have been registerd before via [EventSourcedEntity::cmd].
-#[derive(Debug, Clone)]
-pub struct EntityRef<E, T>
-where
-    E: EventSourced,
-{
-    cmd_in: mpsc::Sender<(BoxedAny, oneshot::Sender<Result<BoxedAny, BoxedAny>>)>,
-    id: E::Id,
-    _e: PhantomData<E>,
-    _t: PhantomData<T>,
-}
+    /// Reply type.
+    type Reply: Debug + Send + 'static;
 
-impl<E, T> EntityRef<E, T>
-where
-    E: EventSourced + 'static,
-{
-    /// The ID of the represented [EventSourced] entity.
-    pub fn id(&self) -> &E::Id {
-        &self.id
-    }
+    /// Error type for rejected (a.k.a. invalid) commands.
+    type Error: StdError + Send + Sync + 'static;
 
-    /// Pass the given command to the represented [EventSourced] entity. The returned value is a
-    /// nested result where the outer one represents technical errors, e.g. problems connecting to
-    /// the event log, and the inner one comes from the command handler, i.e. signals potential
-    /// command rejection.
-    #[instrument(skip(self))]
-    pub async fn handle_cmd<C, X>(
-        &self,
-        cmd: C,
-    ) -> Result<Result<C::Reply, C::Error>, HandleCmdError>
-    where
-        C: Cmd<E> + Debug + Send,
-        T: CoproductSelector<C, X>,
-    {
-        let (result_in, result_out) = oneshot::channel();
-        self.cmd_in
-            .send((Box::new(cmd), result_in))
-            .await
-            .map_err(|_| HandleCmdError("cannot send cmd".to_string()))?;
-        let result = result_out
-            .await
-            .map_err(|_| HandleCmdError("cannot receive cmd handler result".to_string()))?;
-        let result = result
-            .map_err(|error| *error.downcast::<C::Error>().expect("downcast error"))
-            .map(|reply| *reply.downcast::<C::Reply>().expect("downcast reply"));
-        Ok(result)
-    }
-}
+    const TYPE_NAME: &'static str;
 
-/// Extension methods for [EventSourced] entities.
-pub trait EventSourcedExt
-where
-    Self: EventSourced + Sized,
-{
-    /// Create a new [EventSourced] entity for this [EventSourced] implementation.
-    fn entity(self) -> EventSourcedEntity<Self, CNil> {
-        EventSourcedEntity {
-            e: self,
-            cmd_fns: HashMap::new(),
-            _t: PhantomData,
-        }
-    }
-}
+    /// Command handler, returning the to be persisted event or an error.
+    fn handle_cmd(
+        id: &Self::Id,
+        state: &Self::State,
+        cmd: Self::Cmd,
+    ) -> Result<(Self::Evt, Self::ReplyInfo), Self::Error>;
 
-impl<E> EventSourcedExt for E where E: EventSourced {}
+    /// Event handler.
+    fn handle_evt(state: Self::State, evt: Self::Evt) -> Self::State;
 
-/// An [EventSourced] entity which allows for registering `Cmd`s and `spawn`ing.
-pub struct EventSourcedEntity<E, T>
-where
-    E: EventSourced,
-{
-    e: E,
-    cmd_fns: BoxedCmdFns<E>,
-    _t: PhantomData<T>,
+    /// Factory for replies.
+    fn make_reply(info: Self::ReplyInfo, state: &Self::State) -> Self::Reply;
```

Comment:
Is the main reason for the `ReplyInfo` to avoid having to apply the event to the state already in `handle_cmd` to create a reply? Could we solve that without the indirection, e.g. just by expecting users to call `handle_evt` manually another time in `handle_cmd` to create the final reply directly?

Reply:
Because in many cases, creating the reply might not need the full state in which case the indirection here is quite cumbersome.

Reply:
No, users must not call handle_evt manually. But we can pass the event to make_reply.

Reply:
But why have an extra method in the first place? That's the thing that bothers me the most.

Reply:
We could change the command handler to already create a reply which is sent after successful persistence. But then the updated state is not available and even if that is not needed in some cases it will be in other ones. 
Then we could change the command handler to return a closure from the (updated) state to the reply. But how is that less bothersome from a developer experience than the simple make_reply method?

Reply:
It's just about defining things together that belong together, so the reply should be created in the command handler. 

`make_reply` would also work but spread things out more.